### ISSUE TITLE
opam init fails to find zstd when compiling ocaml-base-compiler

### ISSUE THREAD

**Issue by sporkl** [ocaml/ocaml#12835]
I was recently trying for a fresh install of ocaml, and when I ran `opam init -y` it looks like it failed to compile ocaml-base-compiler. When I installed zstd with `brew install zstd`, it worked afterwards.

I tried running it in a clean shell environment, and the same issue occurred as well.

Here's what the output looked like:
```
(base) dv@MacBook-Pro-403 ~ % opam init -y
No configuration file found, using built-in defaults.
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg command on your system.
  - you won't be able to use darcs repositories unless you install the darcs command on your system.


<><> Fetching repository information ><><><><><><><><><><><><><><><><><><><>  🐫 
[default] Initialised

<><> Required setup - please read <><><><><><><><><><><><><><><><><><><><><>  🐫 

  In normal operation, opam only alters files within ~/.opam.

  However, to best integrate with your system, some environment variables
  should be set. If you allow it to, this initialisation step will update
  your zsh configuration by adding the following line to ~/.zshrc:

    [[ ! -r /Users/dv/.opam/opam-init/init.zsh ]] || source /Users/dv/.opam/opam-init/init.zsh  > /dev/null 2> /dev/null

  Otherwise, every time you want to access your opam installation, you will
  need to run:

    eval $(opam env)

  You can always re-run this setup with 'opam init' later.

[WARNING] Shell not updated in non-interactive mode: use --shell-setup

<><> Creating initial switch 'default' (invariant ["ocaml" {>= "4.05.0"}] - initially with ocaml-base-compiler) 

<><> Installing new switch packages <><><><><><><><><><><><><><><><><><><><>  🐫 
Switch invariant: ["ocaml" {>= "4.05.0"}]

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
∗ installed base-bigarray.base
∗ installed base-threads.base
∗ installed base-unix.base
∗ installed ocaml-options-vanilla.1
⬇ retrieved ocaml-base-compiler.5.1.1  (https://opam.ocaml.org/cache)
[ERROR] The compilation of ocaml-base-compiler.5.1.1 failed at "make -j15".

#=== ERROR while compiling ocaml-base-compiler.5.1.1 ==========================#
# context     2.1.5 | macos/x86_64 |  | https://opam.ocaml.org#a4b48f40
# path        ~/.opam/default/.opam-switch/build/ocaml-base-compiler.5.1.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build make -j15
# exit-code   2
# env-file    /var/folders/sw/kx0xn99n7rv5cbyjwdn59z900000gn/T/opam-root-10007/ocaml-base-compiler-10007-cfa107.env
# output-file /var/folders/sw/kx0xn99n7rv5cbyjwdn59z900000gn/T/opam-root-10007/ocaml-base-compiler-10007-cfa107.out
### output ###
# /Applications/Xcode.app/Contents/Developer/usr/bin/make runtime-all
# [...]
#   MKDLL runtime/libcamlrun_shared.so
# /Applications/Xcode.app/Contents/Developer/usr/bin/make -C stdlib \
#         OCAMLRUN='$(ROOTDIR)/runtime/ocamlrun' \
#         CAMLC='$(BOOT_OCAMLC) -use-prims ../runtime/primitives' all
#   OCAMLC camlinternalFormatBasics.cmi
# dyld[15410]: Library not loaded: '@rpath/libzstd.1.dylib'
#   Referenced from: '/Users/dv/.opam/default/.opam-switch/build/ocaml-base-compiler.5.1.1/runtime/ocamlrun'
#   Reason: tried: '/usr/local/lib/libzstd.1.dylib' (no such file), '/usr/lib/libzstd.1.dylib' (no such file)
# make[2]: *** [camlinternalFormatBasics.cmi] Abort trap: 6
# make[1]: *** [coldstart] Error 2
# make: *** [world.opt] Error 2



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><>  🐫 
┌─ The following actions failed
│ λ build ocaml-base-compiler 5.1.1
└─ 
┌─ The following changes have been performed (the rest was aborted)
│ ∗ install base-bigarray         base
│ ∗ install base-threads          base
│ ∗ install base-unix             base
│ ∗ install ocaml-options-vanilla 1
└─ 

<><> ocaml-base-compiler.5.1.1 troubleshooting ><><><><><><><><><><><><><><>  🐫 
=> A failure in the middle of the build may be caused by build parallelism
      (enabled by default).
      Please file a bug report at https://github.com/ocaml/opam-repository/issues
=> You can try installing again including --jobs=1
      to force a sequential build instead.
Switch initialisation failed: clean up? ('n' will leave the switch partially installed) [Y/n] y
[NOTE] Opam has been initialised, but the initial switch creation failed.
       Use 'opam switch create <compiler>' to get started.
```

And here are the results of the configure script (`/var/folders/sw/kx0xn99n7rv5cbyjwdn59z900000gn/T/opam-root-10007/ocaml-base-compiler-10007-cfa107.out`):

```
  CC runtime/sak.o
  MKEXE runtime/sak
/Applications/Xcode.app/Contents/Developer/usr/bin/make coldstart
/Applications/Xcode.app/Contents/Developer/usr/bin/make runtime-all
  GEN runtime/ld.conf
  GEN runtime/caml/opnames.h
  GEN runtime/caml/jumptbl.h
  GEN runtime/build_config.h
  GEN runtime/primitives
  CC runtime/addrmap.b.o
  CC runtime/afl.b.o
  CC runtime/alloc.b.o
  CC runtime/array.b.o
  CC runtime/backtrace.b.o
  CC runtime/bigarray.b.o
  CC runtime/callback.b.o
  CC runtime/codefrag.b.o
  CC runtime/compare.b.o
  CC runtime/custom.b.o
  CC runtime/debugger.b.o
  CC runtime/domain.b.o
  CC runtime/dynlink.b.o
  CC runtime/extern.b.o
  CC runtime/fiber.b.o
  CC runtime/finalise.b.o
  CC runtime/floats.b.o
  CC runtime/gc_ctrl.b.o
  CC runtime/gc_stats.b.o
  CC runtime/globroots.b.o
  CC runtime/hash.b.o
  CC runtime/intern.b.o
  CC runtime/ints.b.o
  CC runtime/io.b.o
  CC runtime/lexing.b.o
  CC runtime/lf_skiplist.b.o
  CC runtime/main.b.o
  CC runtime/major_gc.b.o
  CC runtime/md5.b.o
  CC runtime/memory.b.o
  CC runtime/memprof.b.o
  CC runtime/meta.b.o
  CC runtime/minor_gc.b.o
  CC runtime/misc.b.o
  CC runtime/obj.b.o
  CC runtime/parsing.b.o
  CC runtime/platform.b.o
  CC runtime/printexc.b.o
  CC runtime/prng.b.o
  CC runtime/roots.b.o
  CC runtime/runtime_events.b.o
  CC runtime/shared_heap.b.o
  CC runtime/signals.b.o
  CC runtime/skiplist.b.o
  CC runtime/startup_aux.b.o
  CC runtime/str.b.o
  CC runtime/sync.b.o
  CC runtime/sys.b.o
  CC runtime/unix.b.o
  CC runtime/weak.b.o
  CC runtime/backtrace_byt.b.o
  CC runtime/fail_byt.b.o
  CC runtime/fix_code.b.o
  CC runtime/interp.b.o
  CC runtime/startup_byt.b.o
  CC runtime/zstd.b.o
  CC runtime/addrmap.bd.o
  CC runtime/afl.bd.o
  CC runtime/alloc.bd.o
  CC runtime/array.bd.o
  CC runtime/backtrace.bd.o
  CC runtime/bigarray.bd.o
  CC runtime/callback.bd.o
  CC runtime/codefrag.bd.o
  CC runtime/compare.bd.o
  CC runtime/custom.bd.o
  CC runtime/debugger.bd.o
  CC runtime/domain.bd.o
  CC runtime/dynlink.bd.o
  CC runtime/extern.bd.o
  CC runtime/fiber.bd.o
  CC runtime/finalise.bd.o
  CC runtime/floats.bd.o
  CC runtime/gc_ctrl.bd.o
  CC runtime/gc_stats.bd.o
  CC runtime/globroots.bd.o
  CC runtime/hash.bd.o
  CC runtime/intern.bd.o
  CC runtime/ints.bd.o
  CC runtime/io.bd.o
  CC runtime/lexing.bd.o
  CC runtime/lf_skiplist.bd.o
  CC runtime/main.bd.o
  CC runtime/major_gc.bd.o
  CC runtime/md5.bd.o
  CC runtime/memory.bd.o
  CC runtime/memprof.bd.o
  CC runtime/meta.bd.o
  CC runtime/minor_gc.bd.o
  CC runtime/misc.bd.o
  CC runtime/obj.bd.o
  CC runtime/parsing.bd.o
  CC runtime/platform.bd.o
  CC runtime/printexc.bd.o
  CC runtime/prng.bd.o
  CC runtime/roots.bd.o
  CC runtime/runtime_events.bd.o
  CC runtime/shared_heap.bd.o
  CC runtime/signals.bd.o
  CC runtime/skiplist.bd.o
  CC runtime/startup_aux.bd.o
  CC runtime/str.bd.o
  CC runtime/sync.bd.o
  CC runtime/sys.bd.o
  CC runtime/unix.bd.o
  CC runtime/weak.bd.o
  CC runtime/backtrace_byt.bd.o
  CC runtime/fail_byt.bd.o
  CC runtime/fix_code.bd.o
  CC runtime/interp.bd.o
  CC runtime/startup_byt.bd.o
  CC runtime/zstd.bd.o
  CC runtime/instrtrace.bd.o
  CC runtime/addrmap.bi.o
  CC runtime/afl.bi.o
  CC runtime/alloc.bi.o
  CC runtime/array.bi.o
  CC runtime/backtrace.bi.o
  CC runtime/bigarray.bi.o
  CC runtime/callback.bi.o
  CC runtime/codefrag.bi.o
  CC runtime/compare.bi.o
  CC runtime/custom.bi.o
  CC runtime/debugger.bi.o
  CC runtime/domain.bi.o
  CC runtime/dynlink.bi.o
  CC runtime/extern.bi.o
  CC runtime/fiber.bi.o
  CC runtime/finalise.bi.o
  CC runtime/floats.bi.o
  CC runtime/gc_ctrl.bi.o
  CC runtime/gc_stats.bi.o
  CC runtime/globroots.bi.o
  CC runtime/hash.bi.o
  CC runtime/intern.bi.o
  CC runtime/ints.bi.o
  CC runtime/io.bi.o
  CC runtime/lexing.bi.o
  CC runtime/lf_skiplist.bi.o
  CC runtime/main.bi.o
  CC runtime/major_gc.bi.o
  CC runtime/md5.bi.o
  CC runtime/memory.bi.o
  CC runtime/memprof.bi.o
  CC runtime/meta.bi.o
  CC runtime/minor_gc.bi.o
  CC runtime/misc.bi.o
  CC runtime/obj.bi.o
  CC runtime/parsing.bi.o
  CC runtime/platform.bi.o
  CC runtime/printexc.bi.o
  CC runtime/prng.bi.o
  CC runtime/roots.bi.o
  CC runtime/runtime_events.bi.o
  CC runtime/shared_heap.bi.o
  CC runtime/signals.bi.o
  CC runtime/skiplist.bi.o
  CC runtime/startup_aux.bi.o
  CC runtime/str.bi.o
  CC runtime/sync.bi.o
  CC runtime/sys.bi.o
  CC runtime/unix.bi.o
  CC runtime/weak.bi.o
  CC runtime/backtrace_byt.bi.o
  CC runtime/fail_byt.bi.o
  CC runtime/fix_code.bi.o
  CC runtime/interp.bi.o
  CC runtime/startup_byt.bi.o
  CC runtime/zstd.bi.o
  CC runtime/addrmap.bpic.o
  CC runtime/afl.bpic.o
  CC runtime/alloc.bpic.o
  CC runtime/array.bpic.o
  CC runtime/backtrace.bpic.o
  CC runtime/bigarray.bpic.o
  CC runtime/callback.bpic.o
  CC runtime/codefrag.bpic.o
  CC runtime/compare.bpic.o
  CC runtime/custom.bpic.o
  CC runtime/debugger.bpic.o
  CC runtime/domain.bpic.o
  CC runtime/dynlink.bpic.o
  CC runtime/extern.bpic.o
  CC runtime/fiber.bpic.o
  CC runtime/finalise.bpic.o
  CC runtime/floats.bpic.o
  CC runtime/gc_ctrl.bpic.o
  CC runtime/gc_stats.bpic.o
  CC runtime/globroots.bpic.o
  CC runtime/hash.bpic.o
  CC runtime/intern.bpic.o
  CC runtime/ints.bpic.o
  CC runtime/io.bpic.o
  CC runtime/lexing.bpic.o
  CC runtime/lf_skiplist.bpic.o
  CC runtime/main.bpic.o
  CC runtime/major_gc.bpic.o
  CC runtime/md5.bpic.o
  CC runtime/memory.bpic.o
  CC runtime/memprof.bpic.o
  CC runtime/meta.bpic.o
  CC runtime/minor_gc.bpic.o
  CC runtime/misc.bpic.o
  CC runtime/obj.bpic.o
  CC runtime/parsing.bpic.o
  CC runtime/platform.bpic.o
  CC runtime/printexc.bpic.o
  CC runtime/prng.bpic.o
  CC runtime/roots.bpic.o
  CC runtime/runtime_events.bpic.o
  CC runtime/shared_heap.bpic.o
  CC runtime/signals.bpic.o
  CC runtime/skiplist.bpic.o
  CC runtime/startup_aux.bpic.o
  CC runtime/str.bpic.o
  CC runtime/sync.bpic.o
  CC runtime/sys.bpic.o
  CC runtime/unix.bpic.o
  CC runtime/weak.bpic.o
  CC runtime/backtrace_byt.bpic.o
  CC runtime/fail_byt.bpic.o
  CC runtime/fix_code.bpic.o
  CC runtime/interp.bpic.o
  CC runtime/startup_byt.bpic.o
  CC runtime/zstd.bpic.o
  GEN runtime/prims.c
  MKLIB runtime/libcamlrun.a
  MKLIB runtime/libcamlrund.a
  MKLIB runtime/libcamlruni.a
  CC runtime/prims.o
  MKEXE runtime/ocamlrun
  MKEXE runtime/ocamlrund
  MKEXE runtime/ocamlruni
  MKLIB runtime/libcamlrun_pic.a
  MKDLL runtime/libcamlrun_shared.so
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C stdlib \
	  OCAMLRUN='$(ROOTDIR)/runtime/ocamlrun' \
	  CAMLC='$(BOOT_OCAMLC) -use-prims ../runtime/primitives' all
  OCAMLC camlinternalFormatBasics.cmi
dyld[15410]: Library not loaded: '@rpath/libzstd.1.dylib'
  Referenced from: '/Users/dv/.opam/default/.opam-switch/build/ocaml-base-compiler.5.1.1/runtime/ocamlrun'
  Reason: tried: '/usr/local/lib/libzstd.1.dylib' (no such file), '/usr/lib/libzstd.1.dylib' (no such file)
make[2]: *** [camlinternalFormatBasics.cmi] Abort trap: 6
make[1]: *** [coldstart] Error 2
make: *** [world.opt] Error 2
```

I am on an intel mac running macOS 12.5.1. If I should provide any additional info, please let me know.

Thanks!

**Comment by xavierleroy** [comment]
What does
```
pkg-config --libs libzstd
```
report?

Do you know how ZSTD is installed on your system?  If it is via Homebrew or MacPorts, you could try to upgrade the "zstd" package, or force an uninstall then reinstall it.

Note that OCaml installs just fine if ZSTD is not available.  Things go wrong here because `pkg-config` reports ZSTD as installed, while the corresponding dynamic libraries seem to be missing.

**Comment by sporkl** [comment]
The installation worked fine after I did `brew install zstd`, but prior to that it didn't work. I just found that it looks like I also have zstd installed through mamba, which I'm guessing is the issue. I tried running the installation in a fresh zsh session without mamba loaded, and the issue still happened then. I guess maybe `pkg-config` as able to find it in the mamba location anyways?

Thanks for looking into this!
